File: Sources/HaviKit/Capture/HaviScreenName.swift
import Foundation

/// Pure derivation of a display screen name from a view-controller type name,
/// used by the snapshotter's best-effort auto-detection when the host set no
/// screen via `.haviScreen(_:)` / `Havi.setScreen(_:)`. Kept UIKit-free so the
/// filter runs under `swift test` without a Simulator.
enum HaviScreenName {
    /// Generic container / hosting controllers carry no screen-specific name —
    /// naming a capture "UIHostingController" is worse than falling through to
    /// "unknown", so these resolve to `nil`.
    private static let genericControllers: Set<String> = [
        "UIViewController",
        "UIHostingController",
        "UINavigationController",
        "UITabBarController",
        "UISplitViewController",
        "UIPageViewController",
    ]

    /// The screen name for a fully-qualified controller type name
    /// (`MyApp.ReaderViewController`, `UIHostingController<AnyView>`), or `nil`
    /// when it is a generic container that carries no useful name. The module
    /// prefix and any generic arguments are stripped.
    static func screen(forControllerType typeName: String) -> String? {
        let withoutModule = typeName.split(separator: ".").last.map(String.init) ?? typeName
        let base = withoutModule.split(separator: "<").first.map(String.init) ?? withoutModule
        let trimmed = base.trimmingCharacters(in: .whitespaces)
        guard !trimmed.isEmpty, !genericControllers.contains(trimmed) else { return nil }
        return trimmed
    }
}

File: android/havikit-core/src/main/kotlin/ai/handgemacht/havikit/HaviScreenName.kt
package ai.handgemacht.havikit

/**
 * Pure derivation of a display screen name from a controller/route type name, used
 * by the capture layer's best-effort auto-detection when the host set no screen
 * (wire spec §6.3 precedence). The iOS set covers UIKit generic containers; the
 * Android equivalent adds the framework host classes that carry no screen-specific
 * name. Module prefix and generic arguments are stripped.
 */
public object HaviScreenName {
    private val genericControllers =
        setOf(
            // iOS parity
            "UIViewController",
            "UIHostingController",
            "UINavigationController",
            "UITabBarController",
            "UISplitViewController",
            "UIPageViewController",
            // Android hosts that carry no screen-specific name
            "ComponentActivity",
            "AppCompatActivity",
            "FragmentActivity",
            "NavHostFragment",
        )

    /**
     * The screen name for a fully-qualified type name (`com.app.ReaderActivity`,
     * `UIHostingController<AnyView>`), or null when it is a generic container that
     * carries no useful name.
     */
    public fun screen(typeName: String): String? {
        val withoutModule = typeName.substringAfterLast('.')
        val base = withoutModule.substringBefore('<')
        val trimmed = base.trim()
        if (trimmed.isEmpty() || trimmed in genericControllers) return null
        return trimmed
    }
}
